llama.cpp

```diff
@@ -3750,6 +3750,14 @@ static bool llm_load_tensors(
     model.main_gpu     = main_gpu;
     model.n_gpu_layers = n_gpu_layers;
 
+#ifdef GGML_USE_SYCL
+    if (split_mode == LLAMA_SPLIT_MODE_NONE) {
+        ggml_backend_sycl_set_single_device(main_gpu);
+        //SYCL use device index (0, 1, 2), instead if device id.
+        main_gpu = ggml_backend_sycl_get_device_index(main_gpu);
+    }
+#endif
```

Comment:
Backends should not require calling backend-specific functions for normal usage. Is this `ggml_backend_sycl_set_single_device` function really necessary?

Reply:
I guess the author want to distinguish iGPU and dGPU and offload to dGPU if no device appointed, thus Backends need to query sycl-backend for device list and return the most powerful one. Do you have any suggestion?

Reply:
Could the dGPU always be mapped to the device index zero? That way, it would be used by default in single GPU mode.

Reply:
> Could the dGPU always be mapped to the device index zero? That way, it would be used by default in single GPU mode.

yes, this should be default. But we encountered this https://github.com/ggerganov/llama.cpp/issues/5513, the user reported dGPU mapped to index 3.

Reply:
Ideally, the device indices that the SYCL backend uses for its buffers and backends would be automatically ordered by the most powerful GPUs available in the system, such that the lowest indices are the most powerful GPUs. If this is not possible or desirable, it would still be ok to add a function that returns the list of available GPUs ordered by the most powerful first. Then, in `llama.cpp` we can use that list as a lookup table to convert the device indices used in llama.cpp to the device indices passed to the SYCL backend. This translation between llama.cpp device indices and SYCL backend device indices could be implemented in the `llama_default_buffer_type_offload` function for the buffers, and during the backend instance creation in `llama_new_context_with_model`.

Reply:
@slaren 
Yes, current design idea is as your suggestion:

In SYCL backend, we use device index (from 0, 1, ...) to operate device internally.
SYCL support different type hardware. In a PC, there would be iGPU, dGPU, CPU, FPGA simulator. All are supported by SYCL.
By default, split mode is `layer`. We create a GPUs list for the most powerful GPUs (possible be more than one). If split mode is `none`, the GPU list include one GPU which is set by --main-gpu.

In general, the most powerful GPU will be with device ID 0. But in some cases, it could be other number. And in some case, user want to try with iGPU. So we need to support set main GPU by "--main-gpu" in split mode: `none`.

To set special GPU, user need to set main-gpu with device ID which is shown/printed during SYCL backend init.  When transfer it from llama.cpp to SYCL backend, and we convert the device ID to device index by `ggml_backend_sycl_get_device_index()`. `llama_default_buffer_type_offload` will be inputted `device index` in fact.

Reply:
SYCL backend will create GPU lists with most powerful GPUs in initial.
When llama.cpp provide the parameter: split-mode and main-gpu, SYCL backend will update the GPU list:
1. If split-mode is none, GPU list will be updated to new list include the main-gpu device index only.
2. If split-mode is layer or row, GPU list won't be changed.

Reply:
Better method is to ask user provide gpu list to llama.cpp as parameter.  ggml just create the GPU pool by the parameter, avoid to make mistake by detecting automatically.
It allows to support more feature, like mix GPUs, mix dGPU & iGPU.

If the gpu list include one GPU, the split mode is none in fact.

So, the parameters is changed: 
    from `main-gpu + split-mode`
    to ` gpu-list + split-mode`

Reply:
When using multiple devices with split mode `layer` or `row`, it is possible to exclude some devices by using the `-ts` parameter to set the split of a device to zero. For example, with `-ts 1,0,1` only devices 0 and 2 will be used.

Reply:
SYCL backend has two methods to get the GPUs info:
1. automatically detect the GPUs which are most powerful. It's default behavior in most case, including unit test.
2. according to the parameter: main-gpu comes from llama.cpp. `ggml_backend_sycl_set_single_device` is used in this case only.

It only impacts the GPU list in SYCL backend as global state.
This action will happen before `llama_default_buffer_type_offload`. so, it won't impact next model load process.